### ISSUE TITLE
What is the criteria of graduation from corfxlab?

### ISSUE THREAD

**Issue by codekaizen** [dotnet/corefxlab#134]
[Microsoft.Experimental.Collections](https://www.nuget.org/packages/Microsoft.Experimental.Collections) has a useful implementation of [MultiValueDictionary](http://blogs.msdn.com/b/dotnet/archive/2014/08/05/multidictionary-becomes-multivaluedictionary.aspx), but it now appears abandoned in alpha state. [A discussion on corefx indicates corefxlab is a good place for it](https://github.com/dotnet/corefx/issues/1306), but what is the process to bring it to release from here?

**Comment by KrzysztofCwalina** [comment]
Good question. I have not thought about it. @terrajobst, @richlander, @joshfree, any opinions?

**Comment by Petermarcu** [comment]
@ianhays

**Comment by joshfree** [comment]
@ianhays is going to be open sourcing Microsoft.Expermiental.Collections to /dotnet/corefxlab very soon.

**Comment by ianhays** [comment]
Heya @codekaizen, I wrote this last summer! Now that I'm back at Microsoft I'm planning on moving this along closer to the core repo but there's been some discussion as to where it <I>should</I> go. Our options are to put it either:
- In the dotnet/corefxlab repo.
- In a dev/MultiDictionary branch off of the main dotnet/corefx repo.

The home for the MultiDictionary ultimately depends on how out of the park it is; corefxlab was designed for ideas that were may not make it to the main repo and that are in the more distant future. @terrajobst and I discussed this and were leaning towards just branching it off of master as an addition to System.Collections.Generic, but that decision isn't set yet.

Regardless of where it will be, you can expect to see it somewhere soon :)

**Comment by codekaizen** [comment]
@ianhays - glad to see you back! I'd agree with the sentiment that MultiValueDictionary doesn't seem too far astride of core collection classes, and having it in core would be unremarkable. A branch seems reasonable, but then my question becomes - "what becomes of the branch"? Even if it is merged to master at some point, the current availability of it as a separate package, which has good value for users not running Core, will be lost.

**Comment by ianhays** [comment]
@codekaizen Sorry for the delay; it's been real busy around here! I decided to put it into CoreFXLab for now and just pushed a PR dotnet/corefxlab#370 where we can continue iterating on it until it's ready for CoreFX. For now it is in it's own assembly: System.Collections.Generic.MultiValueDictionary. I tested it with some console apps running on the core framework and it works as expected.

>Even if it is merged to master at some point, the current availability of it as a separate package, which has good value for users not running Core, will be lost.

It's been months but I'm honestly still not sure that I have a good answer to this. In it's current home in CoreFXLab it can easily be built as a desktop NuGet package since it is standalone, but once it is merged into CoreFX it will be jumbled together with the other Collections classes. We could always keep a desktop copy of the nupkg that only contains the MVD and push package updates that way. Although that would be the simplest solution it would require effort to maintain two copies of the same MVD code and subsequently be prone to error or stale versions. Another concern is if the CoreFX MVD evolves to depend on something that doesn't exist in the Desktop framework - what do we do then? How do we prioritize changes and updates to the MultiValueDictionary between Desktop and Core if they diverge and what does that look like? 

To answer the original question: the members of CoreFXLab are primarily focused on eventual inclusion into .NET Core so <b>for now</b> I'm going to keep with that focus and consider the "graduation" to be merging it into CoreFX as a part of System.Collections.Generic.

**Comment by akoeplinger** [comment]
> We could always keep a desktop copy of the nupkg that only contains the MVD and push package updates that way

@ianhays Doesn't the same apply to System.Collections.Immutable? That one is in CoreFx and the nupkg works on Desktop.

**Comment by ianhays** [comment]
Yes, but it is slightly different since Immutable is a standalone library whereas MultiValueDictionary would be built in to System.Collections.Generic. Of course I suppose I could just add a System.Collections.Generic.MultiValueDictionary.Desktop.csproj that only pulls in the MVD files from Open to build the Desktop version and nupkg that only contains the MVD.

**Comment by michaldudak** [comment]
Hi, are there any plans to include the MultiValueDictionary in the BCL or as a (non-prerelease) nuget?

**Comment by YohDeadfall** [comment]
There are no plans to do that. It's archived. See https://github.com/dotnet/corefxlab/pull/2245#issuecomment-386102595.

@ahsonkhan @KrzysztofCwalina This issue probably should be closed.

**Comment by michaldudak** [comment]
Oh well, thanks.

**Comment by joshfree** [comment]
Closing this old issue now that MVD has been archived.

https://github.com/dotnet/corefxlab/blob/master/README.md#archived-projects

https://github.com/dotnet/corefxlab/tree/master/archived_projects/src/System.Collections.Generic.MultiValueDictionary
